### PR TITLE
Boilers

### PR DESCRIPTION
**Pull request by DStrand1**

### CODE REVIEW

File: src/main/java/gregtech/common/metatileentities/multi/MetaTileEntityLargeBoiler.java [hunk 1]
```diff
@@ -440,125 +190,85 @@ private boolean isFireboxPart(IMultiblockPart sourcePart) {
...
-    public <T> T getCapability(Capability<T> capability, EnumFacing side) {
-        T result = super.getCapability(capability, side);
-        if (result != null)
-            return result;
-        if (capability == GregtechCapabilities.CAPABILITY_FUELABLE) {
-            return GregtechCapabilities.CAPABILITY_FUELABLE.cast(this);
+    @Override
+    protected void updateFormedValid() {
+        if (isMufflerFaceFree()) {
+            this.recipeLogic.update();
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/multi/MetaTileEntityLargeBoiler.java:224]: Why is this needed, when the recipe logic update should already be ticking? Could this be what causes the double ticking issue that we were seeing and prevented using the `shouldUpdate` override?

File: src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    private int lastTickSteamOutput;
+    private int excessWater, excessFuel, excessProjectedEU;
+
+    public BoilerRecipeLogic(MetaTileEntityLargeBoiler tileEntity) {
+        super(tileEntity, null);
+    }
+
+    @Override
+    public void update() {
+        if (!isActive() && currentHeat > 0) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java:36]: Should we also lose heat when the boiler is paused with a machine controller or soft hammer?

File: src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        if (currentHeat < getMaximumHeat()) {
+            setHeat(currentHeat + 1);
+        }
+
+        if (++progressTime > maxProgressTime) {
+            completeRecipe();
+        }
+    }
+
+    private int adjustEUtForConfig(int rawEUt) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java:133]: This does not seem to have anything to do with config values, so maybe it should be renamed

File: src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        int throttle = ((MetaTileEntityLargeBoiler) metaTileEntity).getThrottle();
+        return Math.max(25, (int) (rawEUt * (throttle / 100.0)));
+    }
+
+    private int adjustBurnTimeForConfig(int rawBurnTime) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java:138]: This does not seem to have anything to do with config values, so maybe it should be renamed.

File: src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+                // should exit here
+            }
+        }
+*/
+        if (!didStartRecipe) {
+            IItemHandlerModifiable importItems = boiler.getImportItems();
+            for (int i = 0; i < importItems.getSlots(); i++) {
+                ItemStack stack = importItems.getStackInSlot(i);
+                int fuelBurnTime = (int) Math.ceil(ModHandler.getFuelValue(stack));
+                if (fuelBurnTime / 80 > 0) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java:80]: Some comment on why 80 would be nice, even if it is just "This is the chosen divisor"

File: src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+    public int getHeatScaled() {
+        return (int) Math.round(currentHeat / (1.0 * getMaximumHeat()) * 100);
+    }
+
+    public void setHeat(int heat) {
+        if (heat != this.currentHeat && !metaTileEntity.getWorld().isRemote) {
+            writeCustomData(BOILER_HEAT, b -> b.writeVarInt(heat));
+        }
+        this.currentHeat = heat;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java:161]: Should this be moved inside the if?

File: src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+    public int getLastTickSteam() {
+        return lastTickSteamOutput;
+    }
+
+    public void setLastTickSteam(int lastTickSteamOutput) {
+        if (lastTickSteamOutput != this.lastTickSteamOutput && !metaTileEntity.getWorld().isRemote) {
+            writeCustomData(BOILER_LAST_TICK_STEAM, b -> b.writeVarInt(lastTickSteamOutput));
+        }
+        this.lastTickSteamOutput = lastTickSteamOutput;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java:172]: Should this be moved inside the if?

File: src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+    public void invalidate() {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java:175]: invalidate the current heat and the `excess...` variables as well?

File: src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+    @Override
+    protected boolean drawEnergy(int recipeEUt) {
+        GTLog.logger.error("Large Boiler called drawEnergy(), this should not be possible!");
+        return false;
+    }
+
+    @Override
+    protected long getMaxVoltage() {
+        GTLog.logger.error("Large Boiler called getMaxVoltage(), this should not be possible!");
+        return 0;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java:280]: This could return the maximum EU generated per boiler type. I think Tech's generator PR uses this for display, or maybe I am thinking incorrectly.

File: src/main/java/gregtech/common/metatileentities/multi/MetaTileEntityLargeBoiler.java [hunk 10]
```diff
@@ -409,10 +159,10 @@ protected BlockPattern createStructurePattern() {
                 .where('P', states(boilerType.pipeState))
                 .where('X', states(boilerType.fireboxState).setMinGlobalLimited(4)
                         .or(abilities(MultiblockAbility.IMPORT_FLUIDS).setMinGlobalLimited(1))
-                        .or(abilities(MultiblockAbility.IMPORT_ITEMS).setMinGlobalLimited(1)))
+                        .or(abilities(MultiblockAbility.IMPORT_ITEMS).setMinGlobalLimited(1))
+                        .or(autoAbilities())) // muffler, maintenance
```
[the remark this answers is not in the document]
Comment (Syrcan) [reply src/main/java/gregtech/common/metatileentities/multi/MetaTileEntityLargeBoiler.java:166]: I also think that looks weird. But it also makes sort of sense, it's the fuel burning that should emit particles, which is done on the bottom buses/hatches.
But maybe it's better to say screw it and go with what would look better: muffler hatch at the very top. I don't know the answer.

[the remark this answers is not in the document]
Comment (TechLord22) [reply src/main/java/gregtech/common/metatileentities/multi/MetaTileEntityLargeBoiler.java:166]: The fireboxes are what is actually burning the input fuel. It makes sense for the smoke to come out of them. 

A muffler in the main part of the boiler means that the boiler is filling with smoke (instead of steam) and expelling it. Alternatively, it could mean the boiler is expelling (black) steam, which is also weird given the goal is to produce it in a usable manner. Both of these cases don't make much sense.

File: src/main/java/gregtech/common/metatileentities/multi/MetaTileEntityLargeBoiler.java [hunk 11]
```diff
@@ -149,23 +63,28 @@ public MetaTileEntity createMetaTileEntity(MetaTileEntityHolder holder) {
...
+        resetTileAbilities();
+        this.throttlePercentage = 100;
+        this.recipeLogic.invalidate();
+        replaceFireboxAsActive(false);
+    }
+
+    private void initializeAbilities() {
+        this.fluidImportInventory = new FluidTankList(true, getAbilities(MultiblockAbility.IMPORT_FLUIDS));
+        this.itemImportInventory = new ItemHandlerList(getAbilities(MultiblockAbility.IMPORT_ITEMS));
+        this.steamOutputTank = new FluidTankList(true, getAbilities(MultiblockAbility.EXPORT_FLUIDS));
```
[the remark this answers is not in the document]
Comment (DStrand1) [reply src/main/java/gregtech/common/metatileentities/multi/MetaTileEntityLargeBoiler.java:84]: I feel that would limit the boiler too much, as it would lock it to very small hatches in terms of fluid capacity

File: src/main/java/gregtech/integration/jei/GTJeiPlugin.java [hunk 12]
```diff
@@ -150,12 +150,7 @@ public void register(IModRegistry registry) {
                         for (RecipeMap<?> recipeMap : ((IMultipleRecipeMaps) metaTileEntity).getAvailableRecipeMaps()) {
                             registerRecipeMapCatalyst(registry, recipeMap, metaTileEntity);
                         }
-                    } else {
-                        //Special Case here for the processing array
-                        RecipeMap<?> recipeMap = logic.getRecipeMap();
-                        if(recipeMap == null) {
-                            continue;
-                        }
+                    } else if (logic.getRecipeMap() != null) {
```
[the remark this answers is not in the document]
Comment (DStrand1) [reply src/main/java/gregtech/integration/jei/GTJeiPlugin.java:153]: It is needed now for Boilers

File: src/main/java/gregtech/common/metatileentities/multi/MetaTileEntityLargeBoiler.java (hunk 1 above)
Comment (DStrand1) [reply src/main/java/gregtech/common/metatileentities/multi/MetaTileEntityLargeBoiler.java:224]: We need this to force the workable to not run if the muffler is blocked, otherwise muffler would be impactless. There are plenty of ways to do this, but this is the way we have been doing it in other places

File: src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java (hunk 8 above)
Comment (DStrand1) [reply src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java:175]: The idea behind the excess variables is to both not lose potentially gained Steam, but also to not underconsume Water and Fuel on weird burntimes, so I would rather leave these be

File: src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java (hunk 9 above)
Comment (DStrand1) [reply src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java:280]: I'm sure we could do this, but I would rather keep the method names more representative of what they are

**Review by ALongStringOfNumbers — APPROVED** [review]
